toolkit/tools/internal/packagerepo/repoutils/repoutils.go

```diff
@@ -107,3 +71,89 @@ func SaveClonedRepoContents(cloner repocloner.RepoCloner, dstFile string) (err e
 	err = jsonutils.WriteJSONFile(dstFile, repo)
 	return
 }
+
+func removePackageDuplicates(packages []*repocloner.RepoPackage) []*repocloner.RepoPackage {
+	index := 0
+	seen := make(map[string]bool)
+	uniquePackages := make([]*repocloner.RepoPackage, len(packages))
+
+	for _, pkg := range packages {
+		packageID := pkg.ID()
+		if !seen[packageID] {
+			seen[packageID] = true
+			uniquePackages[index] = pkg
+			index++
+		}
+	}
+
+	return uniquePackages[:index]
+}
+
+func removeDownloadedPackages(packages []*repocloner.RepoPackage, cloneDirectory string) []*pkgjson.PackageVer {
+	const packageCondition = "="
+
+	packageVers := make([]*pkgjson.PackageVer, len(packages))
+	index := 0
+	for _, pkg := range packages {
+		pkgVersion := fmt.Sprintf("%s.%s", pkg.Version, pkg.Distribution)
+
+		// Skip packages that are already present, this is expected for the toolchain
```

Comment:
The toolchain RPMs should no longer be present in the chache, they have their own directory now. I think the only reason a package would be here is if it was downloaded as a dependency of another package (but we set that to `false`...). I think if we want to enforce a clean state this could be a hard error.

Reply:
Although since there is a call to `verifyClonedRepoContents()` can we just relax the requirement of a clean checkout here? It could just delete any unexpected RPMs?

Reply:
I've added this function to handle cases of multiple rebuilds with the same summary file. Requiring an empty state seemed overly restrictive and time-consuming. I realized that it's not an error to have pre-existing packages as long as the final set of available packages equaled the ones in the summary file.

I like the idea of removing any unexpected RPMs for builds with summary files. It would be a change in the contract (error out vs silently remove packages) but since this feature hasn't been used by anyone, I think we could get away with it and that's a nice improvement. Having said that, would you mind if I left that change for an undetermined future? Reason: my human limitations. The change seems to be in a stable state now and my mind needs a reset.

Reply:
Quick improvement: I may re-name the function to `filterOutDownloadedPackages`. The word "remove" is hinting at deleting the packages instead of removing them from the list.